File: typescript/src/commands/http-client.ts
/**
 * HTTP client: shared request processing for GET, POST, PUT, PATCH, DELETE.
 *
 * Uses native async fetch() for HTTP requests. Now that the engine is async,
 * we can use fetch() directly without spawning child processes.
 *
 * Mirrors Kotlin's HttpClient.kt and HttpParameters.kt.
 */

import { writeFileSync, mkdirSync } from 'node:fs'
import { dirname, resolve as pathResolve } from 'node:path'
import type { ScriptContext } from '../language/context.js'
import type { JsonValue, JsonObject } from '../language/types.js'
import { isObject, SpecScriptCommandError } from '../language/types.js'
import { parseYamlIfPossible, toDisplayYaml } from '../util/yaml.js'

const HTTP_DEFAULTS_KEY = 'http.defaults'

/**
 * Process an HTTP request from a value form (URL string).
 */
export async function processValueRequest(urlString: string, context: ScriptContext, method: string): Promise<JsonValue | undefined> {
  const encoded = encodePath(urlString)

  let data: JsonObject
  try {
    const uri = new URL(encoded)
    const pathStart = encoded.indexOf(uri.pathname)
    const url = encoded.substring(0, pathStart)
    const path = encoded.substring(pathStart)
    data = url ? { url, path } : { path }
  } catch {
    // Not a full URL — treat entire string as path (needs defaults for url)
    data = { path: encoded }
  }

  return processObjectRequest(data, context, method)
}

/**
 * Process an HTTP request from an object form.
 */
export async function processObjectRequest(data: JsonObject, context: ScriptContext, method: string): Promise<JsonValue | undefined> {
  const defaults = getDefaults(context)
  const merged = mergeWithDefaults({ ...data }, defaults)

  const url = buildUrl(merged)
  const headers = buildHeaders(merged)

  // Add cookies to headers
  addCookieHeader(merged, headers)

  // Add basic auth
  if (merged.username) {
    const username = String(merged.username)
    const password = String(merged.password ?? '')
    headers['Authorization'] = `Basic ${Buffer.from(`${username}:${password}`).toString('base64')}`
  }

  const body = buildBody(merged, headers)
  const saveAs = merged['save as'] as string | undefined

  return executeAsyncRequest(url, method, headers, body, saveAs, context)
}

// --- Defaults management ---

export function storeDefaults(context: ScriptContext, data: JsonObject): void {
  context.session.set(HTTP_DEFAULTS_KEY, data)
}

export function getDefaults(context: ScriptContext): JsonObject | undefined {
  return context.session.get(HTTP_DEFAULTS_KEY) as JsonObject | undefined
}

// --- Parameter merging ---

function mergeWithDefaults(data: JsonObject, defaults: JsonObject | undefined): JsonObject {
  if (!defaults) return data

  for (const [key, value] of Object.entries(defaults)) {
    if (!(key in data)) {
      data[key] = value
    } else if (isObject(data[key]) && isObject(value)) {
      mergeWithDefaults(data[key] as JsonObject, value as JsonObject)
    }
  }
  return data
}

// --- URL building ---

function buildUrl(params: JsonObject): string {
  const host = (params.url as string) ?? ''
  const path = encodePath(params.path as string | undefined)
  return `${host}${path}`
}

function encodePath(path: string | null | undefined): string {
  if (!path) return ''
  return path.replace(/ /g, '+')
}

// --- Headers ---

function buildHeaders(params: JsonObject): Record<string, string> {
  const headers: Record<string, string> = {}

  const paramHeaders = params.headers
  if (isObject(paramHeaders)) {
    for (const [key, value] of Object.entries(paramHeaders)) {
      headers[key] = String(value)
    }
  }

  // Default Content-Type if not set
  if (!Object.keys(headers).some(k => k.toLowerCase() === 'content-type')) {
    headers['Content-Type'] = 'application/json'
  }

  // Default Accept if not set
  if (!Object.keys(headers).some(k => k.toLowerCase() === 'accept')) {
    headers['Accept'] = '*/*'
  }

  return headers
}

// --- Cookies ---

function addCookieHeader(params: JsonObject, headers: Record<string, string>): void {
  const paramCookies = params.cookies
  if (!isObject(paramCookies)) return

  const cookiePairs = Object.entries(paramCookies).map(([k, v]) => `${k}=${v}`)
  if (cookiePairs.length > 0) {
    headers['Cookie'] = cookiePairs.join('; ')
  }
}

// --- Body ---

function buildBody(params: JsonObject, headers: Record<string, string>): string | undefined {
  const body = params.body
  if (body === undefined || body === null) return undefined

  const contentType = Object.entries(headers).find(
    ([k]) => k.toLowerCase() === 'content-type'
  )?.[1] ?? ''

  if (contentType === 'application/x-www-form-urlencoded') {
    if (isObject(body)) {
      return Object.entries(body)
        .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(toDisplayYaml(v as JsonValue))}`)
        .join('&')
    }
  }

  return JSON.stringify(body)
}

// --- Async HTTP execution ---

/**
 * Execute an HTTP request using native async fetch().
 * Now that the engine is async, we can use fetch() directly.
 */
async function executeAsyncRequest(
  url: string,
  method: string,
  headers: Record<string, string>,
  body: string | undefined,
  saveAs: string | undefined,
  context: ScriptContext,
): Promise<JsonValue | undefined> {
  const opts: RequestInit = { method, headers }
  if (body !== undefined) opts.body = body

  let response: Response
  try {
    response = await fetch(url, opts)
  } catch (e) {
    throw new SpecScriptCommandError(
      `HTTP request failed: ${e instanceof Error ? e.message : String(e)}`,
      'connection'
    )
  }

  const bodyText = await response.text()

  // Non-2xx → error
  if (response.status < 200 || response.status >= 300) {
    const data = parseYamlIfPossible(bodyText)
    throw new SpecScriptCommandError(
      'Http request returned an error',
      String(response.status),
      data
    )
  }

  // No content
  const contentLength = response.headers.get('content-length')
  if (contentLength === '0' || (!bodyText && contentLength === null)) {
    return undefined
  }

  // Save to file
  if (saveAs) {
    const filePath = pathResolve(context.workingDir, saveAs)
    mkdirSync(dirname(filePath), { recursive: true })
    writeFileSync(filePath, bodyText, 'utf-8')
    return undefined
  }

  // Parse response body
  return parseYamlIfPossible(bodyText)
}

File: src/main/kotlin/specscript/commands/http/HttpClient.kt
package specscript.commands.http

import io.ktor.http.*
import specscript.language.ScriptContext
import specscript.language.SpecScriptCommandError
import specscript.util.Json
import specscript.util.Yaml
import specscript.util.toDisplayYaml
import tools.jackson.databind.JsonNode
import tools.jackson.databind.node.ObjectNode
import tools.jackson.databind.node.StringNode
import tools.jackson.databind.node.ValueNode
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpRequest
import java.net.http.HttpResponse.BodyHandlers
import java.nio.file.Path

object HttpClient {

    private val client = java.net.http.HttpClient.newHttpClient()

    fun processRequest(data: ValueNode, context: ScriptContext, method: HttpMethod): JsonNode? {

        val uri = URI(encodePath(data.stringValue()))
        val separator = data.stringValue().indexOf(uri.path)
        val parsedData = Json.newObject("path", uri.toString().substring(separator))

        val url = uri.toString().substring(0, separator)
        if (url.isNotEmpty()) {
            parsedData.put("url", url)
        }

        return processRequest(parsedData, context, method)
    }

    fun processRequest(data: ObjectNode, context: ScriptContext, method: HttpMethod): JsonNode? {
        val parameters = HttpParameters.create(data, HttpRequestDefaults.getFrom(context), method)
        return processRequest(parameters)
    }

    private fun processRequest(parameters: HttpParameters): JsonNode? {

        val request = buildRequest(parameters)
        val response = client.send(request, BodyHandlers.ofByteArray())

        return handleResponse(response, parameters)
    }

    private fun buildRequest(parameters: HttpParameters): HttpRequest {
        val builder = HttpRequest.newBuilder()
            .uri(URI.create(parameters.url))
            .method(parameters.method.value, bodyPublisher(parameters))

        headers(builder, parameters)
        cookies(builder, parameters)
        basicAuth(builder, parameters)

        return builder.build()
    }

    private fun bodyPublisher(parameters: HttpParameters): HttpRequest.BodyPublisher {
        val body = parameters.body ?: return HttpRequest.BodyPublishers.noBody()

        val contentType = parameters.headers?.get("Content-Type")?.stringValue()
        if (contentType == ContentType.Application.FormUrlEncoded.toString()) {
            val formData = body.properties().joinToString("&") { entry ->
                "${URLEncoder.encode(entry.key, Charsets.UTF_8)}=${URLEncoder.encode(entry.value.toDisplayYaml(), Charsets.UTF_8)}"
            }
            return HttpRequest.BodyPublishers.ofString(formData)
        }

        return HttpRequest.BodyPublishers.ofString(body.toString())
    }

    private fun headers(builder: HttpRequest.Builder, parameters: HttpParameters) {
        parameters.headers?.properties()?.forEach { header ->
            builder.header(header.key, header.value.stringValue())
        }

        if (parameters.headers?.has("Content-Type") != true) {
            builder.header("Content-Type", ContentType.Application.Json.toString())
        }
        if (parameters.headers?.has("Accept") != true) {
            builder.header("Accept", "*/*")
        }
    }

    private fun cookies(builder: HttpRequest.Builder, parameters: HttpParameters) {
        val cookies = parameters.cookies ?: return
        val cookieHeader = cookies.properties().joinToString("; ") { "${it.key}=${it.value.stringValue()}" }
        builder.header("Cookie", cookieHeader)
    }

    private fun basicAuth(builder: HttpRequest.Builder, parameters: HttpParameters) {
        val username = parameters.username ?: return
        val password = parameters.password ?: ""
        val encoded = java.util.Base64.getEncoder()
            .encodeToString("$username:$password".toByteArray())
        builder.header("Authorization", "Basic $encoded")
    }

    private fun handleResponse(
        response: java.net.http.HttpResponse<ByteArray>,
        parameters: HttpParameters
    ): JsonNode? {

        val statusCode = response.statusCode()

        // Error
        if (statusCode !in 200..299) {
            val data = Yaml.parseIfPossible(String(response.body()))
            val type = statusCode.toString()
            throw SpecScriptCommandError("Http request returned an error", type = type, data = data)
        }

        // No content
        val body = response.body()
        if (body.isEmpty()) return null

        // Save to file
        if (parameters.saveAs != null) {
            Path.of(parameters.saveAs).toFile().writeBytes(body)
            return null
        }

        // Parse body
        return try {
            Yaml.parse(String(body))
        } catch (e: Exception) {
            StringNode(String(body))
        }
    }
}

fun encodePath(path: String?): String {
    return path?.replace(' ', '+') ?: ""
}
